### PR TITLE
Weekly Patch Release v1.4.6 [full merge, no squash]

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?
```bash
gh pr list -s merged -S 'merged:2021-09-01T16:30:00.000Z..2021-09-07T22:30:00.000Z' --json mergedAt,milestone,url,mergeCommit,title --jq 'sort_by(.mergedAt) | reverse | .[] | select((.milestone.title == "v1.4.x") or (.milestone.title == null)) | [.url, .mergeCommit.oid, .title] | join(" ")' --limit 100
```

```
https://github.com/PyTorchLightning/pytorch-lightning/pull/9288 6892d533ea1c743f7e05171846a28e685db85f51 Run plugin closure before `on_before_optimizer_step` [1/2]
https://github.com/PyTorchLightning/pytorch-lightning/pull/9311 d49709e29c6174be8bde7e1edc288180a2173adc Fix DeepSpeed warning CI Test
https://github.com/PyTorchLightning/pytorch-lightning/pull/9319 0135a4bd1ca338ddd1ceedd17c9a91dcd8d8be1f Remove some incorrect comments in ddp.py
https://github.com/PyTorchLightning/pytorch-lightning/pull/9336 98e2f56db090f38a09d5f63202688590702ba15a Clear reference to training loss at the end of train step
https://github.com/PyTorchLightning/pytorch-lightning/pull/9316 9149b649089976d2d723e33fca929bf92f192ff8 [bugfix] Resolve PyTorch Profiling for Manual Optimization
https://github.com/PyTorchLightning/pytorch-lightning/pull/9125 904dde7573c97245b45e477631ece989bd8c01e9 Fix inspection of unspecified args for container hparams
https://github.com/PyTorchLightning/pytorch-lightning/pull/9279 dc3391beaec4e16b08ffe8bf9a05cf8039f8b9e7 Remove deprecation warnings being called for `on_{task}_dataloader`
https://github.com/PyTorchLightning/pytorch-lightning/pull/8877 cf1a589956f86a0cf1a50c0710051eee9b082094 Allow disabling automatic stopping after max_steps or max_epochs
https://github.com/PyTorchLightning/pytorch-lightning/pull/9308 f6d40871bd52ac755a146958513a0a330b813b52 Prevent loss to be moved to the cpu before backward call.
https://github.com/PyTorchLightning/pytorch-lightning/pull/9301 9d0caa6928c28fcf2252c3acdc6fda8570e5adb9 Fix TPU cleaning job
https://github.com/PyTorchLightning/pytorch-lightning/pull/9156 d5ee8d8e3f46f0e5a6789f45d865fb348fd738f3 Disable `{save,check}_on_train_epoch_end` with `check_val_every_n_epoch>1`
https://github.com/PyTorchLightning/pytorch-lightning/pull/9261 f745aa9ce1b8a78b8ef27b939dc1db456837b374 Move tracking epoch end outputs logic to the `EvaluationEpochLoop`
https://github.com/PyTorchLightning/pytorch-lightning/pull/9223 a7461bfc3b98da2314c21603ee457c4b604f4c9a Add missing callbacks to `callbacks.rst`
https://github.com/PyTorchLightning/pytorch-lightning/pull/9232 ead2404aac20658b6ca0d99317bbaabc94f99f87 Added doc strings to base logger file
https://github.com/PyTorchLightning/pytorch-lightning/pull/9231 f0788b3bbc8773543297ee8d8c6d17a679703bb1 scheduled removal of auto_move_data decorator
https://github.com/PyTorchLightning/pytorch-lightning/pull/9267 69cdb79e33de3dc0b19aad4c6fe8c5c9d21d28c4 Add check for uninitialized `_sync_dir` in DDP Plugin to avoid errors during error handling
https://github.com/PyTorchLightning/pytorch-lightning/pull/9289 071ae498083afc131828c982b3fcb62944a751d1 Fix `LightningOptimizer.step` signature
https://github.com/PyTorchLightning/pytorch-lightning/pull/8800 e2ecb8f8591d79e81512cd70d773cb9b4c390132 Allow exporting to onnx when input is tuple
https://github.com/PyTorchLightning/pytorch-lightning/pull/9255 f9994d456cb264f8d66002eae6d7d51bd1ecc94d Update CHANGELOG following patch releases
```

Excluded https://github.com/PyTorchLightning/pytorch-lightning/pull/9231 due to breaking changes and https://github.com/PyTorchLightning/pytorch-lightning/pull/8877 since no bugfix

Also skipping https://github.com/PyTorchLightning/pytorch-lightning/pull/9308 since it's entangled with data fetching.
And https://github.com/PyTorchLightning/pytorch-lightning/pull/9319 because it requires Post LOCALSGD

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]

**Comment by carmocca** [comment]
I think #9288 and #9308 should make it in.

Also, remember you should update the milestones of the PRs with no milestone or that you've decided not to include

**Comment by justusschock** [comment]
i can add #9308 manually. when I tried to cherry pick it, it somehow messed everything up since it included the data fetching. Not sure why...

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -201,10 +201,18 @@ def update_eval_epoch_metrics(self) -> _EVALUATE_OUTPUT:
     """
 
     def on_train_split_start(self, batch_idx: int, split_idx: int, split_batch: Any) -> None:
-        self.trainer._results.extract_batch_size(split_batch)
+        assert self.trainer._results is not None
+        # when the user requests `dataloader_iter`, we can't track the batch_size
+        # and this is left to user responsibility.
+        if isinstance(split_batch, pl.utilities.fetching.DataLoaderIterDataFetcher):
+            self.trainer._results.extract_batch_size(split_batch)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:208]: do we need to remove these lines? It fixes the tests but not entirely sure

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:208]: (it looks like it came in as a conflict merge from this pr: #9336)

**Comment by leezu** [comment]
Would it make sense to revert #9239 as part of 1.4.6? This can trigger "RuntimeError: Only Tensors created explicitly by the user (graph leaves) support the deepcopy protocol at the moment". I'll open an issue about it soon

**Comment by carmocca** [comment]
> I'll open an issue about it soon

You can use #8821 - check the last few comments. A reproduction would be appreciated.

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/early_stopping.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ def __init__(
         check_finite: bool = True,
         stopping_threshold: Optional[float] = None,
         divergence_threshold: Optional[float] = None,
-        check_on_train_epoch_end: bool = True,
+        check_on_train_epoch_end: Optional[bool] = None,
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/early_stopping.py:94]: this change was missing for the failing tests :)

**Comment by awaelchli** [comment]
`test_deepspeed_multigpu_stage_3` test passes but pytest hangs. 🙈 😭 

Somehow deepspeed stage 3 changes made in into this PR even though none of the listed commits have any deepspeed fixes.

**Pull request merged**